src/Libraries/Microsoft.AspNetCore.Telemetry.Middleware/Logging/Internal/HttpLoggingRedactionInterceptor.cs

```diff
@@ -0,0 +1,206 @@
+using Microsoft.Extensions.Telemetry.Logging;
+
+namespace Microsoft.AspNetCore.Telemetry.Http.Logging;
+
+internal sealed class HttpLoggingRedactionInterceptor : IHttpLoggingInterceptor
+{
+    private const string RequestStartTimestamp = "RequestStartTimestamp";
+    private const string Durration = "duration";
+
+    // These three fields are "internal" solely for testing purposes:
+    internal TimeProvider TimeProvider = TimeProvider.System;
+
+    private readonly IncomingPathLoggingMode _requestPathLogMode;
+    private readonly HttpRouteParameterRedactionMode _parameterRedactionMode;
+    private readonly ILogger<HttpLoggingRedactionInterceptor> _logger;
+    private readonly IHttpRouteParser _httpRouteParser;
+    private readonly IHttpRouteFormatter _httpRouteFormatter;
+    private readonly IIncomingHttpRouteUtility _httpRouteUtility;
+    private readonly HeaderReader _requestHeadersReader;
+    private readonly HeaderReader _responseHeadersReader;
+    private readonly string[] _excludePathStartsWith;
+    private readonly IHttpLogEnricher[] _enrichers;
+    private readonly FrozenDictionary<string, DataClassification> _parametersToRedactMap;
+
+    public HttpLoggingRedactionInterceptor(
+        IOptions<LoggingRedactionOptions> options,
+        ILogger<HttpLoggingRedactionInterceptor> logger,
+        IEnumerable<IHttpLogEnricher> httpLogEnrichers,
+        IHttpRouteParser httpRouteParser,
+        IHttpRouteFormatter httpRouteFormatter,
+        IRedactorProvider redactorProvider,
+        IIncomingHttpRouteUtility httpRouteUtility)
+    {
+        var optionsValue = options.Value;
+        _logger = logger;
+        _httpRouteParser = httpRouteParser;
+        _httpRouteFormatter = httpRouteFormatter;
+        _httpRouteUtility = httpRouteUtility;
+
+        _parametersToRedactMap = optionsValue.RouteParameterDataClasses.ToFrozenDictionary(StringComparer.Ordinal, optimizeForReading: true);
+
+        _requestPathLogMode = EnsureRequestPathLoggingModeIsValid(optionsValue.RequestPathLoggingMode);
+        _parameterRedactionMode = optionsValue.RequestPathParameterRedactionMode;
+
+        _requestHeadersReader = new(optionsValue.RequestHeadersDataClasses, redactorProvider);
+        _responseHeadersReader = new(optionsValue.ResponseHeadersDataClasses, redactorProvider);
+
+        _excludePathStartsWith = optionsValue.ExcludePathStartsWith.ToArray();
+
+        _enrichers = httpLogEnrichers.ToArray();
+    }
+
+    public void OnRequest(HttpLoggingContext logContext)
+    {
+        var context = logContext.HttpContext;
+        var request = context.Request;
+        if (ShouldExcludePath(context.Request.Path))
+        {
+            logContext.LoggingFields = HttpLoggingFields.None;
+        }
+        else if (logContext.IsAnyEnabled(HttpLoggingFields.Response))
+        {
+            // We'll need this for the response.
+            // TODO: Should we put a state filed on logContext?
```

Comment:
Using HttpContext for state storage seems fine to me. Any particular problem you are aiming to improve upon with an alternative solution?

Reply:
Just trying to avoid using the Items collection for perf reasons if we already have an object that's going to flow between the two callbacks. However, that state field would have to have similar loose typing so it might not be any better of an experience, and could conflict if there's multiple interceptors.